concent_api/core/payments/report.py

```diff
@@ -0,0 +1,759 @@
+        client_eth_address=requestor_ethereum_address,
+        payment_ts=get_oldest_subtask_creation_date(),
+    )
+    additional_verification_payments_for_provider = service.get_covered_additional_verification_costs(  # pylint: disable=no-value-for-parameter
+        client_eth_address=provider_ethereum_address,
+        payment_ts=get_oldest_subtask_creation_date(),
+    )
+
+    return Payments(
+        regular_payments=regular_payments,
+        settlement_payments=settlement_payments,
+        forced_subtask_payments=forced_subtask_payments,
+        additional_verification_payments_for_requestor=additional_verification_payments_for_requestor,
+        additional_verification_payments_for_provider=additional_verification_payments_for_provider,
+    )
+
+
+def find_forced_subtask_payments_for_subtask_results_settled(
+    subtask_results_settled: PendingResponse,
+    forced_subtask_payments: list,
+) -> List[ForcedSubtaskPaymentEvent]:
+    """ Find forced subtask payments for the same subtask for given SubtaskResultsSettled message. """
+    return [
+        forced_subtask_payment for forced_subtask_payment in forced_subtask_payments
+        if forced_subtask_payment.subtask_id == subtask_results_settled.subtask.subtask_id
+    ]
+
+
+def find_subtask_result_for_additional_verification_payment(
+    additional_verification_payment: CoverAdditionalVerificationEvent,
+    subtask_results_settled: list,
+    subtask_results_rejected: list,
+) -> List[PendingResponse]:
+    """ Find the SubtaskResultsSettled or SubtaskResultsRejected for given additional verification payment. """
+    matched_subtask_results_settled = [
+        results_settled for results_settled in subtask_results_settled
+        if results_settled.subtask.subtask_id == additional_verification_payment.subtask_id
+    ]
+
+    matched_subtask_results_rejected = [
+        results_rejected for results_rejected in subtask_results_rejected
+        if results_rejected.subtask.subtask_id == additional_verification_payment.subtask_id
+    ]
+
+    return matched_subtask_results_settled + matched_subtask_results_rejected
+
+
+def find_settlement_payment_for_force_payment_committed(
+    force_payment_committed_pending_response: PendingResponse,
+    settlement_payments: list,
+) -> List[ForcedPaymentEvent]:
+    """ Find settlement payment for given ForcePaymentCommitted message. """
+    payment_info = force_payment_committed_pending_response.payments.order_by('id').last()
+    return [
+        settlement_payment for settlement_payment in settlement_payments
+        if (
+            int(parse_datetime_to_timestamp(payment_info.payment_ts)) - 10 <=
+            settlement_payment.closure_time <=
+            int(parse_datetime_to_timestamp(payment_info.payment_ts)) + 10
+        )
+    ]
+
+
+def get_errors_and_warnings(payments: Payments, responses: PendingResponses) -> Tuple[List[Error], List[Warnings]]:
```

Comment:
There is very, very lot of low level logic in this function and it is for me really hard to see what is actually happening here. Consider extracting each `for` to separate inner, helper function responsible for low level logic and the names of this helpers will explain high level, business logic.

Reply:
Answered in https://github.com/golemfactory/concent/pull/1073#discussion_r250423558, please keep this discussion in single tread.

Reply:
When I wrote this comment of @kbeker was pending. I haven't seen it, this is the reason of duplication. But i still think it would be better to split... Please continue in  https://github.com/golemfactory/concent/pull/1073#discussion_r250423558